src/registry.rs

```diff
@@ -0,0 +1,62 @@
+use std::sync::atomic::AtomicUsize;
+use std::sync::atomic::Ordering::{Relaxed, Release, SeqCst};
+
+use scope::Namespace;
+
+#[derive(Default, Debug)]
+pub struct Registry {
+    /// The least significant bit is set if the thread is currently pinned. The rest of the bits
+    /// encode the current epoch.
+    state: AtomicUsize,
+}
+
+impl Registry {
+    // FIXME(stjepang): Registrys are stored in a linked list because linked lists are fairly easy
+    // to implement in a lock-free manner. However, traversal is rather slow due to cache misses and
+    // data dependencies. We should experiment with other data structures as well.
+
+    #[inline]
+    pub fn new() -> Self {
+        Self::default()
+    }
+
+    #[inline]
+    pub fn get_state(&self) -> (bool, usize) {
+        let state = self.state.load(Relaxed);
+        ((state & 1) == 1, state & !1)
+    }
+
+    /// Marks the thread as pinned.
+    ///
+    /// Must not be called if the thread is already pinned!
+    #[inline]
+    pub fn set_pinned<N: Namespace>(&self, namespace: N) {
+        let epoch = namespace.epoch().load(Relaxed);
+        let state = epoch | 1;
+
+        // Now we must store `state` into `self.state`. It's important that any succeeding loads
+        // don't get reordered with this store. In order words, this thread's epoch must be fully
+        // announced to other threads. Only then it becomes safe to load from the shared memory.
+        if cfg!(any(target_arch = "x86", target_arch = "x86_64")) {
+            // On x86 architectures we have a choice:
+            // 1. `atomic::fence(SeqCst)`, which compiles to a `mfence` instruction.
+            // 2. `compare_and_swap(_, _, SeqCst)`, which compiles to a `lock cmpxchg` instruction.
+            //
+            // Both instructions have the effect of a full barrier, but the second one seems to be
+            // faster in this particular case.
+            let previous = self.state.load(Relaxed);
+            self.state.compare_and_swap(previous, state, SeqCst);
```

Comment:
@schets So if I got this right, you're suggesting that we do the following?
```rust
self.state.swap(state, SeqCst)
```

This will compile to `xchg` rather than `lock xchg`. The latter has the effect of `fence(SeqCst)` (`mfence`, but I'm not sure about the former. Does `xchg` imply a full fence as well?